### PR TITLE
[MM] Implement Auth Service

### LINKED ISSUE #17: [MM] Implement Auth service
This task involves implementing a service responsible for authorization and authentication within the Mr Market v2 system. The service will protect specified endpoints from unauthorized access, ensuring only authenticated users can access them. The service will also manage two roles: Admin and User, with different levels of access.

### To Do
- Fetch Active Strategies:
  - Authentication & Authorization implementation
  - Authentication should be integrated with Mixin

### Deliverables
- [ ] All tasks completed
- [ ] Unit tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by MWBlocky**
## Description

This Pull Request introduces the implementation of an authorization and authentication service for the Mr Market v2 system. This service is designed to protect specified endpoints from unauthorized access and manage user roles (Admin and User) with different levels of access.

## Summary of changes

- **Implemented `auth` module**.
- **Added role-based access control**: Defined roles (Admin and User) and ensured endpoints are protected based on these roles.
- **Integrated with Mixin**: Authentication is integrated with the Mixin platform for secure user management.
- **Unit tests added**: Verified the correct functionality of authentication and authorization methods.

## Related issues

Fixes #17

### CODE REVIEW

File: packages/mm-backend/src/common/enums/memo.enum.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+export enum TradingType {
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/enums/memo.enum.ts:1]: What's that?

File: packages/mm-backend/src/common/enums/pairs.enum.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+export enum PairsEnum {
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/enums/pairs.enum.ts:1]: What's that

File: packages/mm-backend/src/common/filters/global-exception.filter.ts [hunk 3]
```diff
@@ -1,32 +1,32 @@
...
+      path: httpAdapter.getRequestUrl(ctx.getRequest()),
+    };
 
-        httpAdapter.reply(ctx.getResponse(), responseBody, httpStatus);
-    }
-}
\ No newline at end of file
+    httpAdapter.reply(ctx.getResponse(), responseBody, httpStatus);
+  }
+}
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/filters/global-exception.filter.ts:32]: Isn't there a better way to handle exception HTTP status codes and messages?

File: packages/mm-backend/src/common/interfaces/memo-details.interface.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  symbol: PairsEnum;
+  traceId: string;
+}
+
+export interface MarketMakingMemoDetails {
+  tradingType: TradingType;
+  action: MarketMakingMemoAction;
+  exchangeName: SpotExchange;
+  symbol: PairsEnum;
+  traceId: string;
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/interfaces/memo-details.interface.ts:33]: What's that

File: packages/mm-backend/src/common/utils/auth/jwt.strategy.spec.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import { JwtStrategy } from './jwt.strategy';
+import { ConfigModule, ConfigService } from '@nestjs/config';
+
+describe('JwtStrategy', () => {
+  let strategy: JwtStrategy;
+
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      imports: [
+        ConfigModule.forRoot({
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/utils/auth/jwt.strategy.spec.ts:11]: Add a test to check if a user can access only his orders, strategies etc

File: packages/mm-backend/src/common/utils/auth/jwt.strategy.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    super({
+      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+      ignoreExpiration: false,
+      secretOrKey: configService.get<string>('JWT_SECRET'),
+    });
+  }
+
+  async validate(payload: any) {
+    return {
+      userId: payload.sub,
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/utils/auth/jwt.strategy.ts:18]: Don't we need the mixin client id as well?

File: packages/mm-backend/src/common/utils/memo-decoders.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    spotOrderType: SpotOrderType[spotOrderType],
+    exchangeName: SpotExchange[exchange],
+    destId: destId as PairsEnum,
+    limitPrice: limitPriceOrRefId,
+    refId: refId,
+  };
+};
+
+export const decodeArbitrageMemo = (
+  decodedMemo: string,
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/utils/memo-decoders.ts:39]: What's that?

File: packages/mm-backend/src/common/utils/trading-dispatcher/arbitrage-strategy.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  constructor(private readonly events: EventEmitter2) {}
+
+  execute(decodedMemo: string, snapshot: SafeSnapshot): void {
+    const details = decodeArbitrageMemo(decodedMemo);
+    if (!details) {
+      return;
+    }
+    this.events.emit('arbitrage.create', details, snapshot);
+  }
+}
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/utils/trading-dispatcher/arbitrage-strategy.ts:18]: What's that

File: packages/mm-backend/src/common/utils/trading-dispatcher/market-making-strategy.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  constructor(private readonly events: EventEmitter2) {}
+
+  execute(decodedMemo: string, snapshot: SafeSnapshot): void {
+    const details = decodeMarketMakingMemo(decodedMemo);
+    if (!details) {
+      return;
+    }
+    this.events.emit('market_making.create', details, snapshot);
+  }
+}
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/utils/trading-dispatcher/market-making-strategy.ts:18]: What's that?

File: packages/mm-backend/src/common/utils/trading-dispatcher/spot-strategy.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+  execute(decodedMemo: string, snapshot: SafeSnapshot): void {
+    const details = decodeSpotMemo(decodedMemo);
+    if (!details) {
+      return;
+    }
+    const spotOrderCreateEvent = { ...details, snapshot };
+    this.events.emit('spot.create', spotOrderCreateEvent);
+  }
+}
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/utils/trading-dispatcher/spot-strategy.ts:19]: What's that?

File: packages/mm-backend/src/common/utils/trading-dispatcher/strategy-executor.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    private readonly arbitrageStrategy: ArbitrageStrategy,
+    private readonly marketMakingStrategy: MarketMakingStrategy,
+  ) {
+    this.strategyMap[TradingType.SP] = this.spotStrategy;
+    this.strategyMap[TradingType.AR] = this.arbitrageStrategy;
+    this.strategyMap[TradingType.MM] = this.marketMakingStrategy;
+  }
+
+  executeStrategy(
+    tradingType: string,
```
Comment (piotrswierzy) [inline packages/mm-backend/src/common/utils/trading-dispatcher/strategy-executor.ts:26]: What's that?

File: packages/mm-backend/src/modules/auth/auth.service.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    private readonly mixinGateway: MixinGateway,
+    private configService: ConfigService,
+    private jwtService: JwtService,
+  ) {
+    this.adminPassword = configService.get<string>('ADMIN_PASSWORD');
+  }
+
+  async validateUser(password: string): Promise<string> {
+    if (!this.adminPassword || !password) {
+      throw new UnauthorizedException('Password is required');
```
Comment (piotrswierzy) [inline packages/mm-backend/src/modules/auth/auth.service.ts:25]: This should probably be HTTP Status code: `400`

File: packages/mm-backend/src/modules/auth/auth.service.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+      .digest('hex');
+
+    if (hashedAdminPassword !== password) {
+      throw new UnauthorizedException('Invalid password');
+    }
+
+    const payload = { username: 'admin', roles: ['Admin'], sub: 'admin_id' };
+    return this.jwtService.sign(payload);
+  }
+
```
Comment (piotrswierzy) [inline packages/mm-backend/src/modules/auth/auth.service.ts:39]: Split user and admin auth

File: packages/mm-backend/src/modules/mixin/snapshots/snapshots.service.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
+import { Injectable } from '@nestjs/common';
+import { MixinGateway } from '../../../integrations/mixin.gateway';
+import { Cron } from '@nestjs/schedule';
+import { CustomLogger } from '../../logger/logger.service';
+import { StrategyExecutor } from '../../../common/utils/trading-dispatcher/strategy-executor';
+import { SafeSnapshot } from '@mixin.dev/mixin-node-sdk';
+import { SnapshotsRepository } from './snapshots.repository';
+
+@Injectable()
+export class SnapshotsService {
```
Comment (piotrswierzy) [inline packages/mm-backend/src/modules/mixin/snapshots/snapshots.service.ts:10]: What's that?

File: packages/mm-backend/src/modules/mixin/snapshots/spec/snapshots.service.spec.ts [hunk 15]
```diff
@@ -0,0 +1,159 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { SnapshotsService } from '../snapshots.service';
+import { MixinGateway } from '../../../../integrations/mixin.gateway';
+import { StrategyExecutor } from '../../../../common/utils/trading-dispatcher/strategy-executor';
+import { SnapshotsRepository } from '../snapshots.repository';
+import { CustomLogger } from '../../../logger/logger.service';
+import { SafeSnapshot } from '@mixin.dev/mixin-node-sdk';
+
+describe('SnapshotsService', () => {
```
Comment (piotrswierzy) [inline packages/mm-backend/src/modules/mixin/snapshots/spec/snapshots.service.spec.ts:9]: What's that?

**Review by piotrswierzy — CHANGES_REQUESTED** [review]

File: packages/mm-backend/src/common/entities/snapshots.entity.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
+import { Entity, PrimaryColumn, Column } from 'typeorm';
+
+@Entity()
+export class Snapshot {
```
[the remark this answers is not in the document]
Comment (MWBlocky) [reply packages/mm-backend/src/common/entities/snapshots.entity.ts:4]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/enums/memo.enum.ts (hunk 1 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/enums/memo.enum.ts:1]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/enums/pairs.enum.ts (hunk 2 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/enums/pairs.enum.ts:1]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/filters/global-exception.filter.ts (hunk 3 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/filters/global-exception.filter.ts]: Fixed: 4f17f74b

File: packages/mm-backend/src/common/interfaces/memo-details.interface.ts (hunk 4 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/interfaces/memo-details.interface.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/utils/auth/jwt.strategy.spec.ts (hunk 5 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/utils/auth/jwt.strategy.spec.ts:11]: In 308b9ee5 access_token now includes clientId (mixin user id) which will be used to fetch user resources

File: packages/mm-backend/src/common/utils/auth/jwt.strategy.ts (hunk 6 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/utils/auth/jwt.strategy.ts:18]: fixed: 308b9ee5

File: packages/mm-backend/src/common/utils/memo-decoders.ts (hunk 7 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/utils/memo-decoders.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/utils/trading-dispatcher/arbitrage-strategy.ts (hunk 8 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/utils/trading-dispatcher/arbitrage-strategy.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/utils/trading-dispatcher/market-making-strategy.ts (hunk 9 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/utils/trading-dispatcher/market-making-strategy.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/utils/trading-dispatcher/spot-strategy.ts (hunk 10 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/utils/trading-dispatcher/spot-strategy.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/common/utils/trading-dispatcher/strategy-executor.ts (hunk 11 above)
Comment (MWBlocky) [reply packages/mm-backend/src/common/utils/trading-dispatcher/strategy-executor.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/modules/mixin/snapshots/snapshots.service.ts (hunk 14 above)
Comment (MWBlocky) [reply packages/mm-backend/src/modules/mixin/snapshots/snapshots.service.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/modules/mixin/snapshots/spec/snapshots.service.spec.ts (hunk 15 above)
Comment (MWBlocky) [reply packages/mm-backend/src/modules/mixin/snapshots/spec/snapshots.service.spec.ts]: It's excluded to another branch: Fork-issue_17-snapshot-processing

File: packages/mm-backend/src/modules/auth/auth.service.ts (hunk 12 above)
Comment (MWBlocky) [reply packages/mm-backend/src/modules/auth/auth.service.ts]: Fixed: 3a4da56a

**Review by piotrswierzy — APPROVED** [review]

**Pull request merged**